File: frontend/src/types/skill.ts
export type SkillStatus = "draft" | "review" | "published" | "archived" | "failed";

export type ConversionStatus = "pending" | "processing" | "completed" | "failed";

export type ResourceType = "reference" | "script" | "asset";

export type QualityVerdict = "PASS" | "NEEDS_REVIEW" | "FAIL";

export interface SkillResource {
  id: string;
  skill_id: string;
  version_id: string | null;
  resource_type: string;
  filename: string;
  content_type: string;
  file_size: number;
  extraction_status: string | null;
  created_at: string;
  updated_at: string;
}

export interface SkillVersion {
  id: string;
  skill_id: string;
  version_number: number;
  content: string;
  metadata_json: string;
  change_notes: string;
  is_published: boolean;
  created_by: string;
  created_at: string;
}

export interface SkillListItem {
  id: string;
  name: string;
  description: string;
  product: string;
  status: SkillStatus;
  tags: string;
  quality_score: number | null;
  quality_verdict: QualityVerdict | null;
  structure_check_passed: boolean | null;
  conversion_status: ConversionStatus | null;
  current_version: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  foundry_skill_name: string;
  foundry_sync_status: "none" | "pending" | "synced" | "failed";
  foundry_cloud_version: string;
  foundry_sync_error: string;
}

export interface SourceMaterialInfo {
  id: string;
  name: string;
  product: string;
}

export interface Skill extends SkillListItem {
  therapeutic_area: string;
  compatibility: string;
  metadata_json: string;
  content: string;
  structure_check_details: string;
  quality_details: string;
  conversion_error: string;
  resources: SkillResource[];
  versions: SkillVersion[];
  source_materials: SourceMaterialInfo[];
}

export interface SkillCreate {
  name: string;
  description?: string;
  product?: string;
  therapeutic_area?: string;
  compatibility?: string;
  tags?: string;
  content?: string;
  metadata_json?: string;
}

export interface SkillUpdate extends Partial<SkillCreate> {
  status?: SkillStatus;
}

export interface StructureCheckIssue {
  severity: string;
  dimension: string;
  message: string;
  suggestion: string;
}

export interface StructureCheckResult {
  passed: boolean;
  score: number;
  issues: StructureCheckIssue[];
  content_hash: string;
}

export interface QualityDimension {
  name: string;
  score: number;
  verdict: string;
  strengths: string[];
  improvements: string[];
  critical_issues: string[];
  rationale: string;
}

export type EvaluationStatus = "ai_success" | "ai_unavailable" | "ai_error";

export interface QualityEvaluation {
  overall_score: number;
  overall_verdict: string;
  dimensions: QualityDimension[];
  summary: string;
  top_improvements: string[];
  content_hash: string;
  evaluated_at: string;
  evaluation_status?: EvaluationStatus;
  model_used?: string;
  error_detail?: string;
}

export interface EvaluationCriterion {
  name: string;
  description: string;
}

export interface SkillEvaluationSummary {
  structure_check: {
    passed: boolean | null;
    details: StructureCheckResult | Record<string, never>;
  };
  quality: {
    score: number | null;
    verdict: string | null;
    details: QualityEvaluation | Record<string, never>;
    is_stale: boolean;
    evaluation_status?: EvaluationStatus;
    model_used?: string;
    error_detail?: string;
  };
  evaluation_criteria?: EvaluationCriterion[];
}

export interface PaginatedSkills {
  items: SkillListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SkillFoundryPortalUrlResponse {
  url: string;
  skill_name: string;
  foundry_version: string;
}

File: backend/app/schemas/skill.py
"""Skill request/response schemas (Pydantic v2)."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class SkillCreate(BaseModel):
    """Create a new skill."""

    name: str
    description: str = ""
    product: str = ""
    therapeutic_area: str = ""
    compatibility: str = ""
    tags: str = ""
    content: str = ""
    metadata_json: str = "{}"


class SkillUpdate(BaseModel):
    """Update an existing skill. All fields optional for partial updates."""

    name: str | None = None
    description: str | None = None
    product: str | None = None
    therapeutic_area: str | None = None
    compatibility: str | None = None
    tags: str | None = None
    content: str | None = None
    metadata_json: str | None = None
    status: str | None = None


class SkillResourceOut(BaseModel):
    """Response schema for a skill resource."""

    id: str
    skill_id: str
    version_id: str | None
    resource_type: str
    filename: str
    content_type: str
    file_size: int
    extraction_status: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SkillVersionOut(BaseModel):
    """Response schema for a skill version snapshot."""

    id: str
    skill_id: str
    version_number: int
    content: str
    metadata_json: str
    change_notes: str
    is_published: bool
    created_by: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SkillListOut(BaseModel):
    """Skill list item response (without full content or relationships)."""

    id: str
    name: str
    description: str
    product: str
    status: str
    tags: str
    quality_score: int | None
    quality_verdict: str | None
    structure_check_passed: bool | None
    conversion_status: str | None
    current_version: int
    created_by: str
    created_at: datetime
    updated_at: datetime
    foundry_skill_name: str
    foundry_sync_status: str
    foundry_cloud_version: str
    foundry_sync_error: str

    model_config = ConfigDict(from_attributes=True)


class SourceMaterialInfo(BaseModel):
    """Lightweight reference to a source training material."""

    id: str
    name: str
    product: str

    model_config = ConfigDict(from_attributes=True)


class SkillOut(SkillListOut):
    """Full skill response with content and relationships."""

    therapeutic_area: str
    compatibility: str
    metadata_json: str
    content: str
    structure_check_details: str
    quality_details: str
    conversion_error: str
    resources: list[SkillResourceOut] = []
    versions: list[SkillVersionOut] = []
    source_materials: list[SourceMaterialInfo] = []


class StructureCheckOut(BaseModel):
    """Response schema for L1 structure check result."""

    passed: bool
    score: int
    issues: list[dict]


class QualityEvaluationOut(BaseModel):
    """Response schema for L2 AI quality evaluation result."""

    overall_score: int
    overall_verdict: str
    dimensions: list[dict]
    summary: str
    top_improvements: list[str]


class SkillFoundryPortalUrlResponse(BaseModel):
    """Response schema for the Skill Foundry portal deep-link."""

    url: str
    skill_name: str
    foundry_version: str
